test/e2e/replica_set_x509/replica_set_x509_test.go

```diff
@@ -0,0 +1,222 @@
+)
+
+func TestMain(m *testing.M) {
+	code, err := e2eutil.RunTest(m)
+	if err != nil {
+		fmt.Println(err)
+	}
+	os.Exit(code)
+}
+
+func TestReplicaSetX509(t *testing.T) {
+	resourceName := "mdb-tls"
+	helmArgs := []setup.HelmArg{
+		{Name: "resource.tls.useX509", Value: "true"},
+		{Name: "resource.tls.sampleX509User", Value: "true"},
+	}
+	ctx, testConfig := setup.SetupWithTLS(t, resourceName, helmArgs...)
+	defer ctx.Teardown()
+
+	mdb, _ := e2eutil.NewTestMongoDB(ctx, resourceName, testConfig.Namespace)
+	mdb.Spec.Security.Authentication.Modes = []v1.AuthMode{"X509"}
+	mdb.Spec.Security.TLS = e2eutil.NewTestTLSConfig(false)
+
+	tester, err := FromX509Resource(t, mdb)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Run("User with wrong certificates", func(t *testing.T) {
+		mdb.Spec.Users = []v1.MongoDBUser{
+			getInvalidUser(),
+		}
+		users := mdb.GetAuthUsers()
+
+		t.Run("Create MongoDB Resource", mongodbtests.CreateMongoDBResource(&mdb, ctx))
+		t.Run("Basic tests", mongodbtests.BasicFunctionalityX509(&mdb))
+		t.Run("Agent certificate secrets configured", mongodbtests.AgentX509SecretsExists(&mdb))
+
+		cert, root, dir := createCerts(t, &mdb)
+		defer os.RemoveAll(dir)
+
+		t.Run("Connectivity Fails without certs", tester.ConnectivityFails(WithURI(mongodbtests.GetConnectionStringForUser(mdb, users[0])), WithTls(mdb)))
+		t.Run("Connectivity Fails with invalid certs", tester.ConnectivityFails(WithURI(fmt.Sprintf("%s&tlsCAFile=%s&tlsCertificateKeyFile=%s", mongodbtests.GetConnectionStringForUser(mdb, users[0]), root, cert))))
+	})
+
+	t.Run("User has valid certificate", func(t *testing.T) {
+		t.Run("Update MongoDB Resource", func(t *testing.T) {
+			err := e2eutil.UpdateMongoDBResource(&mdb, func(m *v1.MongoDBCommunity) {
+				m.Spec.Users = []v1.MongoDBUser{getValidUser()}
+			})
+			assert.NoError(t, err)
+		})
+
+		cert, root, dir := createCerts(t, &mdb)
+		defer os.RemoveAll(dir)
+
+		users := mdb.GetAuthUsers()
+
+		t.Run("Basic tests", mongodbtests.BasicFunctionalityX509(&mdb))
+		t.Run("Agent certificate secrets configured", mongodbtests.AgentX509SecretsExists(&mdb))
+		t.Run("Connectivity Succeeds", tester.ConnectivitySucceeds(WithURI(fmt.Sprintf("%s&tlsCAFile=%s&tlsCertificateKeyFile=%s", mongodbtests.GetConnectionStringForUser(mdb, users[0]), root, cert))))
+	})
+
+	t.Run("Rotate agent certificate", func(t *testing.T) {
+		agentCert, err := GetAgentCert(mdb)
+		if err != nil {
+			t.Fatal(err)
+		}
+		initialCertSerialNumber := agentCert.SerialNumber
+
+		initialAgentPem := &corev1.Secret{}
+		err = e2eutil.TestClient.Get(context.TODO(), mdb.AgentCertificatePemSecretNamespacedName(), initialAgentPem)
+		assert.NoError(t, err)
+
+		cert, root, dir := createCerts(t, &mdb)
+		defer os.RemoveAll(dir)
+
+		users := mdb.GetAuthUsers()
+
+		t.Run("Update certificate secret", tlstests.RotateAgentCertificate(&mdb))
+		t.Run("Wait for MongoDB to reach Running Phase after rotating agent cert", mongodbtests.MongoDBReachesRunningPhase(&mdb))
```

Comment:
can we check that mdb is also changing its phase?

Reply:
Ah, a better way to check for a phase change is not to check for different phases since this is racy. What we do in enterprise is to check what the current transitiondate is and whether this has changed. But we can skip this here if we have manually verified it 👍

Reply:
Oh okay, that's good to know, thanks!